Added a search method for the author and title parameters

Hryniv opened the pull request.

Commit:
Add BookSearchParametersDto record class

- Includes fields for titlePart and author.

Commit:
Add Specification interfaces and manager for dynamic query building

- Added SpecificationProviderManager to manage and retrieve SpecificationProvider instances by key.
- Added SpecificationProvider interface to define the contract for providing Specifications based on a key and parameter.
- Added SpecificationBuilder interface for constructing Specifications using BookSearchParametersDto.

Commit:
Add SpecProviderNotFoundException for handling missing specification providers

Commit:
Implement dynamic specifications for Book

- Added AuthorSpecificationProvider and TitleSpecificationProvider to generate Specifications for author and title fields.
- Added BookSpecificationBuilder to construct combined Specifications based on BookSearchParametersDto.
- Added BookSpecificationProviderManager to manage and retrieve appropriate SpecificationProviders.

Commit:
added JpaSpecificationExecutor interface extension to BookRepository

Commit:
Add search method to BookService and to implementation

- Added BookSpecificationBuilder dependency to BookServiceImpl.
- Implemented search method to retrieve books based on BookSearchParametersDto using dynamic Specifications

Commit:
Add searchBooks endpoint to BookController

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,19 @@
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
create constant

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/BookSpecificationBuilder.java:
```diff
@@ -0,0 +1,29 @@
+    public Specification<Book> build(BookSearchParametersDto searchParameters) {
+        Specification<Book> sp = Specification.where(null);
+        if (searchParameters.author() != null) {
+            sp = sp.and(specificationProviderManager.getSpecificationProvider("author")
```
create constants

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java:
```diff
@@ -0,0 +1,19 @@
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
create constant

Elena-Bruyako reviewed (changes requested):
see comments

Commit:
Extracted common key strings into constants in SpecificationProvider implementations and in BookSpecificationBuilder class

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Specification<Book> getSpecification(String param) {
+        return (root, query, criteriaBuilder) -> root.get("author").in(param);
```
use your constant

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Specification<Book> getSpecification(String param) {
+        return (root, query, criteriaBuilder) -> (root.get("title").in(param));
```
same

Elena-Bruyako reviewed (changes requested):
see comments

Commit:
Refactor: Use constants for title and author fields in SpecificationProvider classes

okuzan commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -46,6 +50,15 @@ public BookDto updateBook(Long id, CreateBookRequestDto requestDto) {
+    @Override
+    public List<BookDto> search(BookSearchParametersDto searchParameters) {
+        Specification<Book> spec = bookSpecificationBuilder.build(searchParameters);
+        return bookRepository.findAll(spec)
+                .stream()
+                .map(bookMapper::toDto)
+                .toList();
```
FYI you can declare mapping list to list in mapstruct too with one line

okuzan commented on src/main/java/com/bookstore/repository/book/BookSpecificationProviderManager.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return specificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
change p to provider

Commit:
Refactored BookServiceImpl and BookServiceSpecifictionProviderManager

- Added toDtoList method to BookMapper
- Refactored search and findAll methods in service layer to use BookMapper's toDtoList method for list mapping
- In BookServiceSpecifictionProviderManager changed 'p' to provider

Commit:
corrected the order of imports
